libs/community/langchain_community/vectorstores/sqlitevec.py

```diff
@@ -0,0 +1,242 @@
+    """
+    return struct.pack("%sf" % len(vector), *vector)
+
+
+class SQLiteVec(VectorStore):
+    """SQLite with Vec extension as a vector database.
+
+    To use, you should have the ``sqlite-vec`` python package installed.
+    Example:
+        .. code-block:: python
+            from langchain_community.vectorstores import SQLiteVec
+            from langchain_community.embeddings.openai import OpenAIEmbeddings
+            ...
+    """
+
+    def __init__(
+        self,
+        table: str,
+        connection: Optional[sqlite3.Connection],
+        embedding: Embeddings,
+        db_file: str = "vec.db",
+    ):
+        """Initialize with sqlite client with vss extension."""
+        try:
+            import sqlite_vec  # noqa  # pylint: disable=unused-import
+        except ImportError:
+            raise ImportError(
+                "Could not import sqlite-vec python package. "
+                "Please install it with `pip install sqlite-vec`."
+            )
+
+        if not connection:
+            connection = self.create_connection(db_file)
+
+        if not isinstance(embedding, Embeddings):
+            warnings.warn("embeddings input must be Embeddings object.")
+
+        self._connection = connection
+        self._table = table
+        self._embedding = embedding
+
+        self.create_table_if_not_exists()
+
+    def create_table_if_not_exists(self) -> None:
+        self._connection.execute(
+            f"""
+            CREATE TABLE IF NOT EXISTS {self._table}
+            (
+                rowid INTEGER PRIMARY KEY AUTOINCREMENT,
+                text TEXT,
+                metadata BLOB,
+                text_embedding BLOB
+            )
+            ;
+            """
+        )
+        self._connection.execute(
+            f"""
+            CREATE VIRTUAL TABLE IF NOT EXISTS {self._table}_vec USING vec0(
+                rowid INTEGER PRIMARY KEY,
+                text_embedding float[{self.get_dimensionality()}]
+            )
+            ;
+            """
+        )
+        self._connection.execute(
+            f"""
+                CREATE TRIGGER IF NOT EXISTS embed_text 
+                AFTER INSERT ON {self._table}
+                BEGIN
+                    INSERT INTO {self._table}_vec(rowid, text_embedding)
+                    VALUES (new.rowid, new.text_embedding) 
+                    ;
+                END;
+            """
+        )
+        self._connection.commit()
+
+    def add_texts(
+        self,
+        texts: Iterable[str],
+        metadatas: Optional[List[dict]] = None,
+        **kwargs: Any,
+    ) -> List[str]:
+        """Add more texts to the vectorstore index.
+        Args:
+            texts: Iterable of strings to add to the vectorstore.
+            metadatas: Optional list of metadatas associated with the texts.
+            kwargs: vectorstore specific parameters
+        """
+        max_id = self._connection.execute(
```

Comment:
would a `RETURNING` clause make more sense here, instead of `max_id`? Just to avoid multiple SQL statements, and would avoid subtle bugs with SQLite autoincrement values

Reply:
I wasn't even aware of RETURNING (I copied this impl directly from the sqlite-vss), I'll try that! That also means I can remove the trigger and avoid having two copies of the embedding in the database.

Reply:
I was completely unaware of the RETURNING clause as well. Unfortunately, this does not seem to be working with the sqlite3 `executemany` method as pointed here: https://github.com/python/cpython/issues/100021